src/dlls/mscoree/mscoree.cpp

```diff
@@ -1263,17 +1237,27 @@ HRESULT SetInternalSystemDirectory()
         DWORD len;
 
         // use local buffer for thread safety
-        WCHAR wzSystemDirectory[COUNTOF(g_pSystemDirectory)];
+        WCHAR * wzSystemDirectory = new (nothrow) WCHAR[MAX_LONGPATH+1];
+        if (wzSystemDirectory == NULL)
+        {
+            WCHAR wzSystemDirectoryStack[MAX_LONGPATH+1];
+            wzSystemDirectory = wzSystemDirectoryStack;
+        }
         
-        hr = GetCORSystemDirectoryInternal(wzSystemDirectory, COUNTOF(wzSystemDirectory), &len);
+        hr = GetCORSystemDirectoryInternal(wzSystemDirectory, MAX_LONGPATH+1, &len);
 
         if(FAILED(hr)) {
-            wzSystemDirectory[0] = W('\0');
+            *wzSystemDirectory = W('\0');
             len = 1;
         }
-
+        
         // publish results idempotently with correct memory ordering
         memcpy(g_pSystemDirectory, wzSystemDirectory, len * sizeof(WCHAR));
```

Comment:
... allocate the buffer of just the right size here, fill it in with the path, and then publish it to g_pSystemDirectory. e.g:
```
WCHAR * pSystemDirectory = new (nothrow) WCHAR[len + 1];
... error handling here...
wcscpy_s(pSystemDirectory, len + 1, wzDirectory);
g_pSystemDirectory = pSystemDirectory;
```